crates/ruma-common/src/api/metadata.rs

```diff
@@ -588,6 +570,44 @@ impl MatrixVersion {
         }
     }
 
+    /// Constructor for use by the `metadata!` macro.
+    ///
+    /// Accepts string literals and parses them.
+    #[doc(hidden)]
+    pub const fn from_lit(lit: &'static str) -> Self {
+        use konst::{primitive::parse_u8, string};
+
+        let major: u8;
+        let minor: u8;
+
+        let mut lit_iter = string::split(lit, ".").next();
+
+        {
+            let (checked_first, checked_split) = lit_iter.expect("first iteration always succeeds");
+
+            major = parse_u8(checked_first)
+                .unwrap_or_else(|_| panic!("major version is not a valid number"));
+
+            lit_iter = checked_split.next();
+        }
+
+        match lit_iter {
+            Some((checked_second, checked_split)) => {
+                minor = parse_u8(checked_second)
+                    .unwrap_or_else(|_| panic!("minor version is not a valid number"));
+
+                lit_iter = checked_split.next();
+            }
+            None => panic!("could not find dot to denote second number"),
+        }
+
+        if lit_iter.is_some() {
+            panic!("version literal contains more than one dot")
+        }
+
+        Self::from_parts(major, minor).unwrap_or_else(|_| panic!("not a valid version literal"))
```

Comment:
i do `unwrap_or_else` instead of `expect` as i feel `expect` has a different semantic meaning; "I have checked this invariant" instead of "panic is a valid error behaviour"

we can't return errors either as const fn does not support formatting, so the error would be confusing and obtuse